Modules/IO/JPEG/src/itkJPEGImageIO.cxx

```diff
@@ -207,30 +206,76 @@ JPEGImageIO::Read(void * buffer)
   // prepare to read the bulk data
   jpeg_start_decompress(&cinfo);
 
-  const auto rowbytes = cinfo.output_width * cinfo.output_components;
-  auto *     tempImage = static_cast<JSAMPLE *>(buffer);
-
   auto * volatile row_pointers = new JSAMPROW[cinfo.output_height];
-  for (size_t ui = 0; ui < cinfo.output_height; ++ui)
+
   {
-    row_pointers[ui] = tempImage + rowbytes * ui;
+    const auto rowbytes = cinfo.output_width * this->GetNumberOfComponents();
+    auto *     tempImage = static_cast<JSAMPLE *>(buffer);
+    for (size_t ui = 0; ui < cinfo.output_height; ++ui)
+    {
+      row_pointers[ui] = tempImage + rowbytes * ui;
+    }
   }
 
   // read the bulk data
-  while (cinfo.output_scanline < cinfo.output_height)
+  if (!(m_IsCMYK && m_CMYKtoRGB))
   {
-    if (setjmp(jerr.setjmp_buffer))
+    while (cinfo.output_scanline < cinfo.output_height)
     {
-      itkWarningMacro("JPEG error in the file " << this->GetFileName());
-      jpeg_destroy_decompress(&cinfo);
-      delete[] row_pointers;
-      return;
+      if (setjmp(jerr.setjmp_buffer))
+      {
+        jpeg_destroy_decompress(&cinfo);
+        delete[] row_pointers;
+        itkWarningMacro(<< "JPEG error in the file " << this->GetFileName());
+        return;
+      }
+
+      jpeg_read_scanlines(&cinfo, &row_pointers[cinfo.output_scanline], cinfo.output_height - cinfo.output_scanline);
     }
-    jpeg_read_scanlines(&cinfo, &row_pointers[cinfo.output_scanline], cinfo.output_height - cinfo.output_scanline);
+
+    // finish the decompression step
+    jpeg_finish_decompress(&cinfo);
   }
+  else
+  {
+    JSAMPROW buf1[1];
+    while (cinfo.output_scanline < cinfo.output_height)
+    {
+      auto * volatile buf0 = new JSAMPLE[cinfo.output_width * 4];
```

Comment:
Maybe `cinfo.output_components` could be used instead of `4`? Or add a comment `// CMYK has 4 components`?

Reply:
OK, it is clearer now, i hope. Yes, that `if (!(m_IsCMYK && m_CMYKtoRGB))` was worse readable.